magnet/src/test/java/magnet/inspection/Inspection_VisitInstancesTest.java

```diff
@@ -0,0 +1,148 @@
+package magnet.inspection;
+
+import magnet.Classifier;
+import magnet.Magnet;
+import magnet.Scope;
+import magnet.Scoping;
+import magnet.inspection.events.ObservableScopeVisitor;
+import magnet.inspection.events.OnEnterScope;
+import magnet.inspection.events.OnExitScope;
+import magnet.inspection.events.OnInstance;
+import magnet.internal.InstanceFactory;
+import magnet.internal.InstrumentedScope;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.junit.Before;
+import org.junit.Test;
+
+import static com.google.common.truth.Truth.assertThat;
+import static magnet.inspection.Instance.Provision.BOUND;
+import static magnet.inspection.Instance.Provision.INJECTED;
+
+public class Inspection_VisitInstancesTest {
+
+    private InstrumentedScope scopeA;
+    private InstrumentedScope scopeB;
+
+    private Bound bound = new Bound();
+    private Bound boundA = new Bound();
+    private Bound boundB = new Bound();
+    private String classifierA = "classifierA";
+    private String classifierB = "classifierB";
+
+    private InjectedTopMost injectedTopMost = new InjectedTopMost();
+    private InjectedTopMostFactory injectedTopMostFactory = new InjectedTopMostFactory();
+
+    private InjectedDirect injectedDirect = new InjectedDirect();
+    private InjectedDirectFactory injectedDirectFactory = new InjectedDirectFactory();
+
+    private ObservableScopeVisitor visitor;
+
+    @Before
+    public void before() {
+        scopeA = (InstrumentedScope) new InstrumentedScope(Magnet.createRootScope())
+            .instrumentObjectIntoScope(injectedTopMostFactory, InjectedTopMost.class, injectedTopMost, Classifier.NONE)
+            .instrumentObjectIntoScope(injectedTopMostFactory, InjectedTopMost.class, injectedTopMost, classifierA)
+            .instrumentObjectIntoScope(injectedTopMostFactory, InjectedTopMost.class, injectedTopMost, classifierB)
+            .bind(Bound.class, bound)
+            .bind(Bound.class, boundA, classifierA);
+
+        scopeB = (InstrumentedScope) ((InstrumentedScope) scopeA.createSubscope())
+            .instrumentObjectIntoScope(injectedDirectFactory, InjectedDirect.class, injectedDirect, Classifier.NONE)
+            .bind(Bound.class, boundB, classifierB);
+    }
+
+    @Test
+    public void visitAllScopes() {
+        visitor = new ObservableScopeVisitor();
+        scopeA.accept(visitor, Integer.MAX_VALUE);
+
+        assertThat(visitor.visited.get(0)).isEqualTo(new OnEnterScope(scopeA.scope, null));
```

Comment:
Nit: Before accessing individual items of a list in a test I usually check for the expected total length of the list first, to rule out small test modifications. Otherwise you usually hunt down an issue like "why isn't this reported as that?!" for some time until you notice that just an intermediate item was added or removed.

Reply:
Agree. Fixed.